### DISCUSSION TITLE
Calling JS from Rust

### DISCUSSION THREAD

**Discussion by rdwnsjjd** [tauri-apps/tauri#3018] [Q&A] [answered]
How can I call js from rust?

I think [this](https://tauri.studio/en/docs/guides/command/#returning-data) just responses to a js call.

**Comment by amrbashir** [comment, marked as answer]
> How can I call js from rust?

Use events, which is the preferred way or assign your js functions to the global `window` object and use `window.eval("window['function_name']()")` in rust.

> I think [this](https://tauri.studio/en/docs/guides/command/#returning-data) just responses to a js call.

That's correct.

**Reply by FabianLars** [reply]
https://docs.rs/tauri/1.0.0-beta.8/tauri/window/struct.Window.html#method.eval

**Reply by rdwnsjjd** [reply]
oh in rust. ok tnx

**Reply by rdwnsjjd** [reply]
the events just emitted on start, right?? how can I use them whenever I want??
I want to emit an event when I finished a function called by js for expamle.

**Reply by FabianLars** [reply]
> Oh in rust

Yes, of course. That's what you asked for :D

> How to emit events

https://tauri.studio/en/docs/guides/events/

**Reply by rdwnsjjd** [reply]
I read it. but it emits events in setup.
I don't underestand why is helpful for me? I want to emit events from MY custom function

**Reply by FabianLars** [reply]
You can emit events from everywhere where you have access to a Window or App(Handle) struct, that includes commands: https://tauri.studio/en/docs/guides/command/#accessing-the-window-in-commands

**Reply by ZodiacWind** [reply]
> You can emit events from everywhere where you have access to a Window or App(Handle) struct, that includes commands: https://tauri.studio/en/docs/guides/command/#accessing-the-window-in-commands

I have encountered a smilar problem: I want to emit events when clicking the menu, for example:

```
        .on_menu_event(|event| {
            match event.menu_item_id() {
                "trigger" => {
                    // how to emit events here?
                    // or how should I get app(Handle)?
                }
                _ => {}
            }
        })
```

What should I do here? Thanks!

**Reply by FabianLars** [reply]
@ZodiacWind 
```rs
        .on_menu_event(|event| {
            let window = event.window();
            match event.menu_item_id() {
                "trigger" => {
                    window.emit();
                }
                _ => {}
            }
        })
```
That's for the global menu event handler on the `tauri::Builder` struct. For handlers registered on a specific window ([this](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.MenuEvent.html)) you can just clone the window handle you're using to register the handler.

**Reply by ZodiacWind** [reply]
It works! 
> For handlers registered on a specific window ([this](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.MenuEvent.html)) you can just clone the window handle you're using to register the handler.

I'm not sure if I've got your point. For example, the menu event is emitted on a Window "Emitter", then I should clone the window handler of "Emitter" instead of the global one? Is that right?

If I use `window.emit_all()` instead of `window.emit()`, then the JS parts of all windows can receive the event. Is that right?

One more question, what's the difference between AppHandle and Window Handle? How could I get AppHandle in the above scenario (i.e.  on_menu_event) ? And how should I choose between them two?

**Reply by ZodiacWind** [reply]
I just noticed that I can get AppHanlder when I use ` .on_system_tray_event()`. Maybe when dealing with app-wide features, I can get AppHandle, otherwise I cannot get it.

**Reply by FabianLars** [reply]
@ZodiacWind 
> I'm not sure if I've got your point

There are 2 different .on_menu_functions, see here: https://docs.rs/tauri/1.0.0-rc.6/tauri/async_runtime/?search=.on_menu_event. My code example is for the first one (the Builder one) and what i mentioned in the text was for the other one.

> If I use window.emit_all() instead of window.emit(), then the JS parts of all windows can receive the event. Is that right?

Yes.

> One more question, what's the difference between AppHandle and Window Handle?

Well one is a handle the whole App and the other to a specific Window. They cover different use cases and have different apis: [AppHandle](https://docs.rs/tauri/1.0.0-rc.6/tauri/struct.AppHandle.html) - [Window](https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.Window.html).

For emitting events you can use both since they both offer the functions from the [Manager](https://docs.rs/tauri/1.0.0-rc.6/tauri/trait.Manager.html#method.emit_to) trait. The Window additionally has a special function to emit just to that window (the same functionality as "emit_to" but without the need to specify the window label).

Listening to events is a little bit more constrained. You can again use the Manager trait on both the Window and AppHandle to listen to global events. For window specific events you need to use functions on the Window struct tho: <https://docs.rs/tauri/1.0.0-rc.6/tauri/window/struct.Window.html#method.listen>

> How could I get AppHandle in the above scenario (i.e. on_menu_event)?

```rs
let handle = event.window().app_handle();
```

> And how should I choose between them two?

Well, you choose the one you need or the one you have easier access to 🤷

**Reply by ZodiacWind** [reply]
Got it. Thank you sooo much for your prompt reply. It's really informative and helpful!

**Reply by Hellager** [reply]
@FabianLars @ZodiacWind 
Sorry for disturbing, but once i have tried to clone the AppHandle or Window to a state managed struct, the whole program will be stuck. How should i do if want rust to send events activly like i asked here https://github.com/tauri-apps/tauri/discussions/5111
